src/app/clusters/actions-server/actions-server.cpp

```diff
@@ -0,0 +1,471 @@
+{
+    ChipLogProgress(Zcl, "ActionsServer: OnStateChanged");
+
+    // Generate StateChanged event
+    EventNumber eventNumber;
+    Events::StateChanged::Type event{ aActionId, aInvokeId, aActionState };
+
+    if (CHIP_NO_ERROR != LogEvent(event, aEndpoint, eventNumber))
+    {
+        ChipLogError(Zcl, "ActionsServer: Failed to generate OnStateChanged event");
+    }
+}
+
+void ActionsServer::OnActionFailed(EndpointId aEndpoint, uint16_t aActionId, uint32_t aInvokeId, ActionStateEnum aActionState,
+                                   ActionErrorEnum aActionError)
+{
+    ChipLogProgress(Zcl, "ActionsServer: OnActionFailed");
+
+    // Generate ActionFailed event
+    EventNumber eventNumber;
+    Events::ActionFailed::Type event{ aActionId, aInvokeId, aActionState, aActionError };
+
+    if (CHIP_NO_ERROR != LogEvent(event, aEndpoint, eventNumber))
+    {
+        ChipLogError(Zcl, "ActionsServer: Failed to generate OnActionFailed event");
+    }
+}
+
+CHIP_ERROR ActionsServer::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    VerifyOrDie(aPath.mClusterId == Actions::Id);
+
+    switch (aPath.mAttributeId)
+    {
+    case ActionList::Id: {
+        ReturnErrorOnFailure(aEncoder.EncodeList(
+            [this, aPath](const auto & encoder) -> CHIP_ERROR { return this->ReadActionListAttribute(aPath, encoder); }));
+        return CHIP_NO_ERROR;
+    }
+    case EndpointLists::Id: {
+        ReturnErrorOnFailure(aEncoder.EncodeList(
+            [this, aPath](const auto & encoder) -> CHIP_ERROR { return this->ReadEndpointListAttribute(aPath, encoder); }));
+        return CHIP_NO_ERROR;
+    }
+    default:
+        break;
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ActionsServer::ReadActionListAttribute(const ConcreteReadAttributePath & aPath,
+                                                  const AttributeValueEncoder::ListEncodeHelper & aEncoder)
+{
+    Delegate * delegate = GetDelegate(aPath.mEndpointId);
+    if (delegate == nullptr)
+    {
+        ChipLogError(Zcl, "Actions delegate is null!!!");
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+
+    for (uint16_t i = 0; i < kMaxActionListLength; i++)
+    {
+        ActionStructStorage action;
+        CHIP_ERROR err = delegate->ReadActionAtIndex(i, action);
+
+        if (err == CHIP_ERROR_PROVIDER_LIST_EXHAUSTED)
+        {
+            return CHIP_NO_ERROR;
+        }
+
+        ReturnErrorOnFailure(aEncoder.Encode(action));
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ActionsServer::ReadEndpointListAttribute(const ConcreteReadAttributePath & aPath,
+                                                    const AttributeValueEncoder::ListEncodeHelper & aEncoder)
+{
+    Delegate * delegate = GetDelegate(aPath.mEndpointId);
+    if (delegate == nullptr)
+    {
+        ChipLogError(Zcl, "Actions delegate is null!!!");
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
```

Comment:
There should be a method that is checked for success that does the logging so this boilerplate is not everywhere.

Reply:
Done